ibc-core/ics24-host/types/src/identifiers/chain_id.rs

```diff
@@ -116,6 +115,173 @@ impl ChainId {
+                        write!(formatter, "expected one of: {:?}", &FIELDS)
+                    }
+
+                    fn visit_str<E>(self, value: &str) -> Result<Field, E>
+                    where
+                        E: Error,
+                    {
+                        match value {
+                            "id" => Ok(Field::Id),
+                            "revisionnumber" | "revision_number" => Ok(Field::RevisionNumber),
+                            _ => Err(Error::unknown_field(value, FIELDS)),
+                        }
+                    }
+                }
+
+                deserializer.deserialize_identifier(FieldVisitor)
+            }
+        }
+
+        struct ChainIdVisitor;
+
+        impl<'de> Visitor<'de> for ChainIdVisitor {
+            type Value = ChainId;
+
+            fn expecting(&self, formatter: &mut Formatter<'_>) -> fmt::Result {
+                formatter.write_str("struct ChainId")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: Error,
+            {
+                match ChainId::from_str(v) {
+                    Ok(chain_id) => Ok(chain_id),
+                    Err(e) => Err(Error::custom(format_args!(
+                        "invalid chain ID `{}`: `{}`",
+                        v, e
+                    ))),
+                }
+            }
+
+            fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                let mut id = None;
+                let mut revision_number = None;
+
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        Field::Id => {
+                            if id.is_some() {
+                                return Err(Error::duplicate_field("id"));
+                            }
+
+                            let next_value = map.next_value::<&str>()?;
+
+                            match parse_chain_id_string(next_value) {
+                                Ok((chain_name, rn)) => {
+                                    validate_prefix_length(chain_name, 1, 64).map_err(|_| {
+                                        Error::custom("invalid chain ID prefix length")
+                                    })?;
+
+                                    id = Some(next_value);
+                                    revision_number = Some(rn);
+                                }
+                                _ => {
+                                    validate_identifier_length(next_value, 1, 64).map_err(
+                                        |_| Error::custom("invalid chain ID identifier length"),
+                                    )?;
+
+                                    id = Some(next_value);
+                                    revision_number = Some(0);
+                                }
+                            }
```

Comment:
`validate_identifier_chars(next_value)` should be performed before this match statement. This part can be simplified by using `ChainId::from_str` instead, as follows:

```suggestion
                            let chain_id =
                                match ChainId::from_str(next_value).map_err(Error::custom) {
                                    Ok(chain_id) => chain_id,
                                    Err(e) => return Err(e),
                                };
                            id = Some(chain_id.id);
                            revision_number = Some(chain_id.revision_number)
```

Good to have a test case for a `ChainId` with invalid chars e.g. `#[case(r#"{"id":"/foo-42","revision_number":"0"}"#)]` must fail.

Reply:
[50aa991](https://github.com/cosmos/ibc-rs/pull/1013/commits/50aa991cc362da69cd749300fabcdd01c9df6c20)